public/app/features/commandPalette/useMatches.ts

```diff
@@ -0,0 +1,256 @@
+   * Represents the commandScore matchiness value which we use
+   * in addition to the explicitly set `action.priority` to
+   * calculate a more fine tuned fuzzy search.
+   */
+  score: number;
+};
+
+function useInternalMatches(filtered: ActionImpl[], search: string): Match[] {
+  const ufuzzy = useUfuzzy();
+
+  const value = React.useMemo(
+    () => ({
+      filtered,
+      search,
+    }),
+    [filtered, search]
+  );
+
+  const { filtered: throttledFiltered, search: throttledSearch } = useThrottledValue(value);
+
+  return React.useMemo(() => {
+    if (throttledSearch.trim() === '') {
+      return throttledFiltered.map((action) => ({ score: 0, action }));
+    }
+
+    const haystack = throttledFiltered.map((action) =>
+      [action.name, action.keywords, action.subtitle].join(' ').toLowerCase()
+    );
+
+    const results: Match[] = [];
+
+    // If the search term is too long, then just do a simple substring search.
+    // We don't expect users to actually hit this frequently, but want to prevent browser hangs
+    if (throttledSearch.length > FUZZY_SEARCH_LIMIT) {
+      const query = throttledSearch.toLowerCase();
+
+      for (let haystackIndex = 0; haystackIndex < haystack.length; haystackIndex++) {
+        const haystackItem = haystack[haystackIndex];
+
+        // Use the position of the match as a stand-in for score
+        const substringPosition = haystackItem.toLowerCase().indexOf(query);
+
+        if (substringPosition > -1) {
+          const score = haystack.length - substringPosition;
+          const action = throttledFiltered[haystackIndex];
+          results.push({ score, action });
+        }
+      }
+    } else {
+      const allMatchedIndexes = new Set<number>();
+
+      const queryWords = ufuzzy.split(throttledSearch);
+      const queryPermutations = uFuzzy.permute(queryWords).map((terms) => terms.join(' '));
+
+      for (const permutedSearchTerm of queryPermutations) {
+        const indexes = ufuzzy.filter(haystack, permutedSearchTerm);
+        const info = ufuzzy.info(indexes, haystack, permutedSearchTerm);
+        const order = ufuzzy.sort(info, haystack, permutedSearchTerm);
+
+        for (let orderIndex = 0; orderIndex < order.length; orderIndex++) {
+          const actionIndex = order[orderIndex];
+
+          if (!allMatchedIndexes.has(actionIndex)) {
+            allMatchedIndexes.add(actionIndex);
+            const score = order.length - orderIndex;
+            const action = throttledFiltered[info.idx[actionIndex]];
+            results.push({ score, action });
+          }
+        }
+      }
+    }
+
+    return results;
+  }, [throttledFiltered, throttledSearch, ufuzzy]);
+}
+
+const FUZZY_SEARCH_LIMIT = 25;
+
+function useUfuzzy(): uFuzzy {
+  const ref = React.useRef<uFuzzy>();
+
+  if (!ref.current) {
+    ref.current = new uFuzzy({
+      intraMode: 1,
+      intraIns: 1,
+      intraSub: 1,
+      intraTrn: 1,
+      intraDel: 1,
+    });
+  }
+
+  return ref.current;
+}
```

Comment:
This section is basically the only part that I've written/changed. Everything else is copy-pasta